src/chat.rs

```diff
@@ -673,30 +674,24 @@ pub fn msgtype_has_file(msgtype: Viewtype) -> bool {
 
 fn prepare_msg_common(context: &Context, chat_id: u32, msg: &mut Message) -> Result<u32, Error> {
     msg.id = 0;
-
     if msg.type_0 == Viewtype::Text {
         // the caller should check if the message text is empty
     } else if msgtype_has_file(msg.type_0) {
-        let path_filename = msg.param.get(Param::File);
-
-        ensure!(
-            path_filename.is_some(),
-            "Attachment missing for message of type #{}.",
-            msg.type_0
-        );
-
-        let mut path_filename = path_filename.unwrap().to_string();
-
-        if msg.state == MessageState::OutPreparing && !dc_is_blobdir_path(context, &path_filename) {
-            bail!("Files must be created in the blob-directory.");
-        }
-
-        ensure!(
-            dc_make_rel_and_copy(context, &mut path_filename),
-            "Failed to copy"
-        );
-
-        msg.param.set(Param::File, &path_filename);
+        let param = msg.param.get(Param::File).ok_or_else(|| {
+            format_err!("Attachment missing for message of type #{}.", msg.type_0)
+        })?;
+        let file = ParamsFile::from_param(context, param)?;
```

Comment:
maybe a new msg.param.get_file() could already return a ParamsFile?

Reply:
I like that idea!

Reply:
I ended up adding `Params::get_file()`, `Params::get_path()` and `Params::get_blob()`, significantly simplifying all the callers.  Thanks for the hint!